contracts/VaultV3.vy

```diff
@@ -551,13 +551,88 @@ def _max_deposit(receiver: address) -> uint256:
 
 @view
 @internal
-def _max_redeem(owner: address) -> uint256:
-    return self.balance_of[owner]
+def _max_withdraw(
+    owner: address,
+    max_loss: uint256,
+    strategies: DynArray[address, MAX_QUEUE]
+) -> uint256:
+    """
+    @dev Returns the max amount of `asset` an `owner` can withdraw.
 
-@view
-@internal
-def _max_withdraw(owner: address) -> uint256:
-    return self._convert_to_assets(self.balance_of[owner], Rounding.ROUND_DOWN)
+    This will do a full simulation of the withdraw in order to determine
+    how much is currently liquid and if the `max_loss` would allow for the 
+    tx to not revert.
+
+    This will track any expected loss to check if the tx will revert, but
+    not account for it in the amount returned since it is unrealised and 
+    therefore will not be accounted for in the conversion rates.
+
+    i.e. If we have 100 debt and 10 of unrealised loss, the max we can get
+    out is 90, but a user of the vault will need to call withdraw with 100
+    in order to get the full 90 out.
+    """
+    # Get the max amount for the owner if fully liquid.
+    max_assets: uint256 = self._convert_to_assets(self.balance_of[owner], Rounding.ROUND_DOWN)
+    
+    # See if we have enough idle to service the withdraw.
+    current_idle: uint256 = self.total_idle
+    if max_assets > current_idle:
+        # Amount left that we need.
+        needed: uint256 = unsafe_sub(max_assets, current_idle)
+        # Track how much we can pull.
+        have: uint256 = current_idle
+        loss: uint256 = 0
+
+        # If no queue was passed use the default one.
+        _strategies: DynArray[address, MAX_QUEUE] = strategies
+        if len(_strategies) == 0:
+            _strategies = self.default_queue
+
+        for strategy in _strategies:
+            # Can't use an invalid strategy.
+            assert self.strategies[strategy].activation != 0, "inactive strategy"
+
+            # Get the maximum amount the vault can withdraw from the strategy.
+            max_withdraw: uint256 = min(needed, self.strategies[strategy].current_debt)
+
+            # Get any unrealised loss for the strategy.
+            unrealised_loss: uint256 = self._assess_share_of_unrealised_losses(strategy, max_withdraw)
+
+            # See if any limit is enforced by the strategy.
+            strategy_limit: uint256 = IStrategy(strategy).maxWithdraw(self)
+
+            # Adjust accordingly if there is a max withdraw limit.
+            if strategy_limit < max_withdraw - unrealised_loss:
+                # lower unrealised loss to the proportion to the limit.
+                unrealised_loss = unrealised_loss * strategy_limit / max_withdraw
+                # Still count the unrealised loss as withdrawable.
+                max_withdraw = strategy_limit + unrealised_loss
```

Comment:
So this was an issue i was unsure how to handle. But I believe we have to count it as "withdrawable" since it is unrealized in order to comply with the standard.

I wrote the reasoning in the comments. https://github.com/Schlagonia/yearn-vaults-v3/blob/limits/contracts/VaultV3.vy#L566-L572

Because the conversion rates the vault uses dont yet know about the loss the amounts need to not adjust for it.

The maxRedeem would also be screwed up if we didnt account for the unrealized loss as withdrawable.

Its not an ideal way to do it but im pretty sure it has to be done this way to comply to the standard.

For a withdraw to be able to take on unrealized losses a custom non-zero 'max_loss' parameter must be used, so i would expect any integrators who pass in a non zero value to know its then possible to receive less than requested.